webidls/Pipeline.webidl

```diff
@@ -0,0 +1,176 @@
+  "complete", // The document and all sub-resources have finished loading. The state indicates that the load event has been fired. (same as document.readyState)
+}
+
+enum SaveType {
+  "HTMLOnly", // Save only the HTML of the page.
+  "HTMLComplete", // Save complete-html page.
+  "MHTML", // Save complete-html page as MHTML.
+}
+
+dictionary PipelineError { // crash reports, DNS/TCP errors, … Not HTTP error (see Pipeline.HTTPResponse).
+  USVString finalURL;
+  // Chrome error list: https://cs.chromium.org/chromium/src/net/base/net_error_list.h
+  unsigned short code;
+  DOMString description; // Human readable
+  optional DOMString report; // Backtrace for panics
+}
+
+enum SaveRenderingStrategy {
+  "default",
+  "displaylist",
+  "texture",
+}
+
+interface Pipeline {
+
+  Promise<void> setPipelineHandler(PipelineHandler handler);
+
+  readonly attribute USVString url; // Happens during redirects for example. FIXME: should that be finalURL ? How often would that change?
+  readonly attribute DOMString title;
+  readonly attribute unsigned short HTTPResponse;
+
+  // Used to replace mozbrowserconnected, mozbrowserloadstart, mozbrowserloadend
+  // Use performance for time stamps.
+  readonly attribute PipelineState? pipelineState;
+  readonly attribute Performance performance; // See Performance.webidl and PerformanceTiming.webidl // FIXME: should probably be getPerformance(). Necessary to implement https://developer.mozilla.org/en-US/Add-ons/WebExtensions/API/history/HistoryItem
+  readonly attribute PipelineError pipelineError; // pipeline-state-changed -> pipelineState == "error" || "crash". // FIXME: is "crash" necessary? // FIXME: so generic…
+
+  readonly attribute FrozenArray<USVString> icons;
+  readonly attribute FrozenArray<MetaTag> metas;
+  readonly attribute String? hoveredLink;
+  readonly attribute ConnectionSecurity connectionSecurity;
+
+  readonly attribute boolean isFrozen;
+  readonly attribute boolean isVisible;
+  readonly attribute unsigned float devicePixelRatio;
+
+  Promise<unsigned float> setDevicePixelRatio(unsigned float ratio); // Will fail for non-top level pipelines
+
+  // Prerendering a document. HTTP + layout, no JS. Lot of things to consider. See:
+  // https://bugzilla.mozilla.org/show_bug.cgi?id=730101
+  // Will be rendered once Browser.navigateToPipeline() is called.
+  // Created via Browser.createPrerenderingPipeline()
+  readonly attribute boolean isPrerender;
+  Promise<void> cancelPrerenderAndRequestDestruction();
+
+  void stopLoading();
+
+  // Will destroy that pipeline and create a new one
+  void reload();
+  void clearCacheAndReload();
+
+  Promise<LoadData> buildLoadData();
+
+  // We want to be able to render frozen pipeline, so we need
+  // a way to save the rendering.
+  // FIXME: do we really want to let the client handle that?
+  readonly attribute SaveRenderingStrategy saveRenderingStrategy;
+  Promise<void> setSaveRenderingStrategy(SaveRenderingStrategy);
```

Comment:
Hmm… I was thinking we might want to hint the engine about this. Like, if we know that the user will soon go back, we want to restore the pipeline and set a saving strategy that is CPU-friendly.